cmd/kcp-catalog/bind/catalogentry/cmd.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	if !strings.HasPrefix(b.CatalogEntryRef, "root") {
+		return fmt.Errorf("fully qualified reference to workspace where Catalog Entry exists is required. The format is `root:<ws>:<catalogentry>`")
+	}
+
+	return b.Options.Validate()
+}
+
+// Run creates an apibinding for the user.
+func (b *BindOptions) Run(ctx context.Context) error {
+	config, err := b.ClientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	kcpclient, err := newKCPClusterClient(config)
+	if err != nil {
+		return err
+	}
+
+	_, currentClusterName, err := pluginhelpers.ParseClusterURL(config.Host)
+	if err != nil {
+		return err
+	}
+
+	// get the entry refered in the command.
+	path, entryName := logicalcluster.New(b.CatalogEntryRef).Split()
+
+	client, err := newCatalogClient(config)
+	if err != nil {
+		return err
+	}
+
+	// TODO: figure out if this usage of controller-runtime client can be used.
+	entry := catalogv1alpha1.CatalogEntry{}
+	// TODO: context shouldn't embed path directly, figure out if the host URL will be correct.
+	err = client.Get(logicalcluster.WithCluster(ctx, path), types.NamespacedName{Name: entryName}, &entry)
+	if err != nil {
+		return err
+	}
+
+	apibindings := []apisv1alpha1.APIBinding{}
+	for _, ref := range entry.Spec.Exports {
+		// check if ref is valid. Skip if invalid by logging error.
+		if ref.Workspace.Path == "" || ref.Workspace.ExportName == "" {
+			klog.Infof("invalid reference %q/%q", ref.Workspace.Path, ref.Workspace.ExportName)
+			continue
+		}
+
+		apibinding := &apisv1alpha1.APIBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: ref.Workspace.ExportName,
+			},
+			Spec: apisv1alpha1.APIBindingSpec{
+				Reference: ref,
+			},
+		}
+
+		apibindings = append(apibindings, *apibinding)
+	}
+
+	// Create bindings to the target workspace
+	for _, binding := range apibindings {
+		_, err := kcpclient.Cluster(currentClusterName).ApisV1alpha1().APIBindings().Create(ctx, &binding, metav1.CreateOptions{})
+		if err != nil {
+			// If an APIBinding already exists, intentionally not updating it since we would not like reset AcceptablePermissionClaims.
+			klog.Infof("Failed to create API binding %s: %w", binding.Name, err)
+			continue
+		}
+
+		if err := wait.PollImmediate(time.Millisecond*500, time.Second*5, func() (done bool, err error) {
+			createdBinding, err := kcpclient.Cluster(currentClusterName).ApisV1alpha1().APIBindings().Get(ctx, binding.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+			if createdBinding.Status.Phase == apisv1alpha1.APIBindingPhaseBound {
+				return true, nil
+			}
+			return false, nil
+		}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func newKCPClusterClient(config *rest.Config) (*kcpclient.Cluster, error) {
+	clusterConfig := rest.CopyConfig(config)
+	u, err := url.Parse(config.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	u.Path = ""
+	clusterConfig.Host = u.String()
+	clusterConfig.UserAgent = rest.DefaultKubernetesUserAgent()
```

Comment:
Need to use a non-default UA

Reply:
My bad, I didn't push the late changes. Its all updated to using controller-runtime clients now.